README
AzeOP
===========
Version: 0.1.0
Date: 2015-12-08

AzeOP is a cool web application.

You are looking at a sample CallBackery application. It comes complete
with a classic "configure - make - install" setup.

Prerequisite
------------

Get a copy of the Qooxdoo SDK from www.qooxdoo.org.

 mkdir ~/src
 cd ~/src
 wget http://downloads.sourceforge.net/qooxdoo/qooxdoo-5.0-sdk.zip
 unzip qooxdoo-5.0-sdk.zip

Setup
-----
Now got back to your app source directory start building

 ./configure --prefix=$HOME/opt/aze_op --with-qooxdoo-sdk-dir=$HOME/src/qooxdoo-5.0-sdk
 make

Configure will check if all requirements are met and give
hints on how to fix the situation if something is missing.

Any mising perl modules will be downloaded and built.

Development
-----------

While developing the application it is conveniant to NOT have to install it
before runnning. You can actually serve the Qooxdoo source directly
using the built-in Mojo webserver.

  cd frontend
  make source
  cd ../bin
  ./aze_op-source-mode.sh

You can now connect to the CallBackery app with your web browser.

If you need any additional perl modules, write their names into the PERL_MODULES
file and run ./bootstrap.

Installation
------------

To install the application, just run

   make install

You can now run aze_op.pl in reverse proxy mode.

   cd $HOME/opt/aze_op/bin
   ./aze_op.pl prefork

Packaging
---------

Before release, make sure to update CHANGES, VERSION and run ./bootstrap

You can also package the application as a nice tar.gz file, it will contain
a mini copy of cpan, so that all perl modules can be rebuilt at the
destination.  If you want to make sure that your project builds with perl
5.10.1, make sure to set the PERL environment variable to a perl 5.10.1 interpreter,
make sure to delete any PERL5LIB environment variable, remove the thirdparty
directory from your source tree and configure again.  Now all modules to build your
project with any perl from 5.10.1 on upward will be included in the distribution.

   make dist

Enjoy!

Tobias Oetiker <tobi@oetiker.ch>


CHANGES
0.0.0 2015-15-11/21/15  Tobias Oetiker (tobi@oetiker.ch)

- started project


lib/AzeOP.pm
package AzeOP;

use Mojo::Base 'CallBackery';

=head1 NAME

AzeOP - the application class

=head1 SYNOPSIS

 use Mojolicious::Commands;
 Mojolicious::Commands->start_app('AzeOP');

=head1 DESCRIPTION

Configure the mojolicious engine to run our application logic

=cut

=head1 ATTRIBUTES

AzeOP has all the attributes of L<CallBackery> plus:

=cut

=head2 config

use our own plugin directory and our own configuration file:

=cut

has config => sub {
    my $self = shift;
    my $config = $self->SUPER::config(@_);
    $config->file($ENV{AzeOP_CONFIG} || $self->home->rel_file('etc/aze_op.cfg'));
    unshift @{$config->pluginPath}, 'AzeOP::GuiPlugin';
    return $config;
};

has database => sub {
    my $self = shift;
    my $database = $self->SUPER::database(@_);
    $database->sql->migrations
        ->name('AzeOPBaseDB')
        ->from_data(__PACKAGE__,'appdb.sql')
        ->migrate;
    return $database;
};

1;

=head1 COPYRIGHT

Copyright (c) 2015 by OETIKER+PARTNER AG. All rights reserved.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=cut

__DATA__

@@ appdb.sql

-- 1 up

-- time tracking (arbeits zeit erfassung)
CREATE TABLE aze (
    aze_id          INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    aze_cbuser      INTEGER NOT NULL REFERENCES cbuser(cbuser_id),
    aze_start       DATETIME NOT NULL,
    aze_duration    INTEGER, -- seconds
    aze_note        TEXT,
    aze_lastmod     TEXT NOT NULL
);

-- absence tracking (ferien / krankheit)
CREATE TABLE abs (
    abs_id          INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    abs_cbuser      INTEGER NOT NULL REFERENCES cbuser(cbuser_id),
    abs_start       DATE NOT NULL,
    abs_duration    INEGER NOT NULL, -- days
    abs_note        TEXT
    abs_lastmod     TEXT NOT NULL
);

-- public holidays (oeffentliche feiertage)
CREATE TABLE pho (
    pho_id          INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    pho_date        DATE NOT NULL,
    pho_title       TEXT NOT NULL,
    pho_lastmod     TEXT NOT NULL
);

-- employment status (anstellungsgrad)
CREATE TABLE ems (
    ems_id          INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    ems_cbuser      INTEGER NOT NULL REFERENCES cbuser(cbuser_id),
    ems_start       DATE NOT NULL,
    ems_percent     INTEGER NOT NULL,
    ems_note        TEXT,
    ems_lastmod     TEXT NOT NULL
);

-- add an extra right for people who can edit

INSERT INTO cbright (cbright_key,cbright_label)
    VALUES ('record','Arbeitszeit Erfassen');

-- 1 down

DROP TABLE aze;
DROP TABLE abs;
DROP TABLE pho;
DROP TABLE ems;


lib/AzeOP/GuiPlugin/Report.pm
package AzeOP::GuiPlugin::Report;
use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable';
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use POSIX qw(strftime);

=head1 NAME

AzeOP::GuiPlugin::Aze - Aze Table

=head1 SYNOPSIS

 use AzeOP::GuiPlugin::Aze;

=head1 DESCRIPTION

The Aze Table Gui.

=cut

=head1 METHODS

All the methods of L<CallBackery::GuiPlugin::AbstractTable> plus:

=cut

has formCfg => sub {
    my $self = shift;
    my @reports = (
        {title => trm('Tag'), key => 'day'},
        {title => trm('Woche'), key => 'week'},
        {title => trm('Monat'), key => 'month'},
        {title => trm('Jahr'), key => 'year'},
    );
    my %reports;
    for my $r (@reports) {
        $reports{$r->{key}} = $r->{title};
    }

    return [
        {
            key => 'report_sel',
            label => trm('Report'),
            widget => 'selectBox',
            cfg => {
                required => $self->true,
                structure => \@reports,
            },
            validator => sub {
                my $value = shift;
                return trm("Unbekannter Report")
                    if not $reports{$value};
                return undef;
            },
        }
    ]
};

=head2 tableCfg

=cut

has tableCfg => sub {
    my $self = shift;
    return [
        {
            label => trm('Periode'),
            type => 'str',
            width => '1*',
            key => 'period',
            sortable => $self->true,
        },
        $self->user->may('admin') ? (
        {
            label => trm('User'),
            type => 'str',
            width => '2*',
            key => 'login',
            sortable => $self->true,
        },
        ):(),
        {
            label => trm('Total'),
            type => 'str',
            width => '2*',
            key => 'total',
            sortable => $self->true,
        },
        {
            label => trm('Soll'),
            type => 'str',
            width => '2*',
            key => 'required',
            sortable => $self->true,
        },
        {
            label => trm('Differenz'),
            type => 'str',
            width => '2*',
            key => 'difference',
            sortable => $self->true,
        }
    ]
};


=head2 actionCfg

Only users who can write get any actions presented.

=cut

has strftimeFormat => sub {
    return {
        day => '%Y-%m-%d',
        week => '%Y.%W',
        month => '%Y-%m',
        year => '%Y'
    };
};

sub getTableRowCount {
    my $self = shift;
    my $args = shift;
    my $db = $self->user->mojoSqlDb;
    my $strftime = $db->dbh->quote($self->strftimeFormat->{$args->{formData}{report_sel} || 'day'});
    my $WHERE = '';
    if (! $self->user->may('admin')){
        $WHERE = 'WHERE aze_cbuser = ' . $db->dbh->quote($self->user->userId);
    }

    return ($db->dbh->selectrow_array(<<"SQL_END"))[0];
SELECT SUM(x)
  FROM (
    SELECT 1 AS x
    FROM aze
    $WHERE
    GROUP BY strftime($strftime,aze_start), aze_cbuser
  );
SQL_END
}

sub getTableData {
    my $self = shift;
    my $args = shift;
    my $db = $self->user->mojoSqlDb;
    my $strftime = $db->dbh->quote($self->strftimeFormat->{$args->{formData}{report_sel} || 'day'});
    my $SORT ='ORDER by period DESC';
    if ($args->{sortColumn}){
        $SORT = 'ORDER BY '.$db->dbh->quote_identifier($args->{sortColumn});
        $SORT .= $args->{sortDesc} ? ' DESC' : ' ASC';
    }
    my $WHERE = '';
    if (! $self->user->may('admin')){
        $WHERE = 'WHERE aze_cbuser = ' . $db->dbh->quote($self->user->userId);
    }

    return $db->dbh->selectall_arrayref(<<"SQL_END",{Slice => {}}, $args->{lastRow}-$args->{firstRow}+1,$args->{firstRow});
SELECT
    strftime($strftime,aze_start) AS period,
    cbuser_login as login,
    printf('%.1f',sum(ifnull(aze_duration,0))/3600.0) as total,
    sum(NULL) AS required,
    sum(NULL) AS difference
FROM aze JOIN cbuser ON aze_cbuser = cbuser_id
    $WHERE
    GROUP BY period
    $SORT
LIMIT ? OFFSET ?
SQL_END
}

1;
__END__

=head1 COPYRIGHT

Copyright (c) 2015 by Tobias Oetiker. All rights reserved.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2015-34-08/27/15 to 0.0 first version

=cut


lib/AzeOP/GuiPlugin/AzeForm.pm
package AzeOP::GuiPlugin::AzeForm;
use Mojo::Base 'CallBackery::GuiPlugin::AbstractForm';
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use POSIX qw(strftime);

=head1 NAME

AzeOP::GuiPlugin::AzeForm - Time Tracker Edit Form

=head1 SYNOPSIS

 use AzeOP::GuiPlugin::AzeForm;

=head1 DESCRIPTION

The Time Tracker Edit Form

=cut

=head1 METHODS

All the methods of L<CallBackery::GuiPlugin::AbstractForm> plus:

=cut

=head2 formCfg

Returns a Configuration Structure for the TimeTracker Entry Form.

=cut


has formCfg => sub {
    my $self = shift;
    my $db = $self->user->db;
    my $cbUsers = $db->getMap('cbuser','login');
    my %cbUsers;
    
    for (@$cbUsers){
        $cbUsers{$_->{key}} = $_->{title};
    }

    return [
        $self->config->{type} eq 'edit' ? {
            key => 'aze_id',
            label => trm('Id'),
            widget => 'text',
            set => {
                readOnly => $self->true,
            },
        }:(),
        $self->user->may('admin') ? (
            {
                key => 'aze_cbuser',
                label => trm('User'),
                widget => 'selectBox',
                cfg => {
                    required => $self->true,
                    structure => $cbUsers
                },
                validator => sub {
                    my $value = shift;
                    return trm("Ungültiger user")
                        if not $cbUsers{$value};
                    return undef;
                },
            }
        ) :(),
        {
            key => 'aze_start',
            label => trm('Start'),
            widget => 'text',
            set => {
                required => $self->true,
                placeholder => 'YYYY-MM-DD HH:MM:SS',
            },
            validator => sub {
                my $value = shift;
                if ($value !~ /^\d{4}-\d\d-\d\d \d\d:\d\d:\d\d$/) {
                    return "Startzeit als YYYY-MM-DD HH:MM:SS erwartet";
                }
                return undef;
            }
        },
        {
            key => 'aze_duration',
            label => trm('Dauer'),
            widget => 'text',
            set => {
                placeholder => 'HH:MM',
            },
            validator => sub {
                my $value = shift;
                if ($value !~ /^((\d?\d:\d\d)|(\d+(\.\d+)?))$/){
                    return "Dauer als HH:MM erwartet";
                }
                return undef;
            }
        },
        {
            key => 'aze_note',
            label => trm('Note'),
            widget => 'textArea',
            set => {
                placeholder => 'some extra information about this entry',
            }
        },
    ];
};

has actionCfg => sub {
    my $self = shift;
    my $type = $self->config->{type} // 'new';

    my $handler = sub {
        my $args = shift;
        if ($args->{aze_duration}){
            if ($args->{aze_duration} =~ /:/){
                my @duration = split /:/, $args->{aze_duration};
                $args->{aze_duration} = $duration[2]*3600+$duration[1]*60*$duration[0];
            }
            else {
                $args->{aze_duration} *= 3600;
            }
        }
        my @userMatch = ();
        if (! $self->user->may('admin')){
            $args->{aze_cbuser} = $self->user->userId;
            @userMatch = 'aze_cbuser' => $self->user->userId;
        }
        $args->{aze_lastmod} = strftime('%Y-%m-%d %H:%M:%S',localtime(time))
                        . ' - Edit  by '. $self->user->userInfo->{cbuser_login};

        my @fields = qw(start duration note);

        if ($type eq 'new'){
            push @fields, 'cbuser';
        }

        my $db = $self->user->db;

        my $id = $db->updateOrInsertData('aze',{
            map { $_ => $args->{'aze_'.$_} } @fields
        },$args->{aze_id} ? { id => int($args->{aze_id}), @userMatch } : (@userMatch));
        return {
            action => 'dataSaved'
        };
    };
    return [
        {
            label => $type eq 'edit'
               ? trm('Speichern')
               : trm('Hinzufügen'),
            action => 'submit',
            key => 'save',
            handler => $handler
        }
    ];
};

has grammar => sub {
    my $self = shift;
    $self->mergeGrammar(
        $self->SUPER::grammar,
        {
            _doc => "Tree Node Configuration",
            _vars => [ qw(type) ],
            type => {
                _doc => 'type of form to show: edit, add',
                _re => '(edit|add)'
            },
        },
    );
};

sub getAllFieldValues {
    my $self = shift;
    my $args = shift;
    return {} if $self->config->{type} ne 'edit';
    my $id = $args->{selection}{aze_id};
    return {} unless $id;

    my $db = $self->user->db;
    my $data = $db->fetchRow('aze',{id => $id});
    if ($data->{aze_duration}){
        $data->{aze_duration} = sprintf("%.1f", $data->{aze_duration}/3600);
    }
    return $data;
}

has checkAccess => sub {
    my $self = shift;
    return $self->user->may('admin');
};

1;
__END__

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2015-15-11/21/15 to 0.0 first version

=cut


lib/AzeOP/GuiPlugin/Aze.pm
package AzeOP::GuiPlugin::Aze;
use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable';
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use POSIX qw(strftime);

=head1 NAME

AzeOP::GuiPlugin::Aze - Aze Table

=head1 SYNOPSIS

 use AzeOP::GuiPlugin::Aze;

=head1 DESCRIPTION

The Aze Table Gui.

=cut

my $getLastEntry = sub {
    my $self = shift;
    my $db = $self->user->mojoSqlDb;
    return $db->dbh->selectrow_hashref(<<"SQL_END",{},$self->user->userId);
        SELECT aze_id,aze_start,strftime('%s',aze_start,'utc') AS aze_epoch, aze_duration
        FROM aze
        WHERE
            aze_cbuser = ?
            AND date(aze_start,'localtime') = date('now','localtime')
        ORDER BY aze_start DESC
        LIMIT 1
SQL_END
};

=head1 METHODS

All the methods of L<CallBackery::GuiPlugin::AbstractTable> plus:

=cut

has formCfg => sub {
    my $self = shift;
    return [] unless $self->user->may('record');
    return [
        {
            key => 'tt_today',
            label => trm('Stopuhr'),
            widget => 'text',
            set => {
                readOnly => $self->true,
                width => 200
            },
            getter => sub {
                my $rec = $self->$getLastEntry();
                if ($rec ){
                    if (not $rec->{aze_duration}) {
                        return strftime('%H:%M:%S',gmtime(time - $rec->{aze_epoch}));
                    }
                }
                return trm('mit [Zeit Erfassen] starten')
            }
        }
    ]
};

=head2 tableCfg

=cut

has tableCfg => sub {
    my $self = shift;
    return [
        {
            label => trm('Id'),
            type => 'str',
            width => '1*',
            key => 'aze_id',
            sortable => $self->true,
        },
        $self->user->may('admin') ? (
        {
            label => trm('User'),
            type => 'str',
            width => '2*',
            key => 'cbuser_login',
            sortable => $self->true,
        },
        ):(),
        {
            label => trm('Datum'),
            type => 'str',
            width => '3*',
            key => 'c_date',
            sortable => $self->true,
        },
        {
            label => trm('Begin'),
            type => 'str',
            width => '2*',
            key => 'c_begin',
            sortable => $self->true,
        },
        {
            label => trm('Ende'),
            type => 'str',
            width => '2*',
            key => 'c_end',
            sortable => $self->true,
        },
        {
            label => trm('Total'),
            type => 'str',
            width => '2*',
            key => 'c_total',
            sortable => $self->true,
        },
        {
            label => trm('Bearbeitet am'),
            type => 'str',
            width => '6*',
            key => 'aze_lastmod',
            sortable => $self->true,
        },
        {
            label => trm('Bemerkung'),
            type => 'str',
            width => '2*',
            key => 'aze_note',
            sortable => $self->true,
        }
    ]
};


=head2 actionCfg

Only users who can write get any actions presented.

=cut

has actionCfg => sub {
    my $self = shift;
    my $recordTime = sub {
        my $rec = $self->$getLastEntry();
        my $db = $self->user->db;
        if ( ref $rec eq 'HASH' and not $rec->{aze_duration} ){
            $db->updateOrInsertData('aze',{
                    duration => time - int($rec->{aze_epoch}),
                    lastmod => strftime('%Y-%m-%d %H:%M:%S',localtime(time))
                        . ' - RecEnd by ' . $self->user->userInfo->{cbuser_login}
                },
                {
                    id => int($rec->{aze_id})
                }
            );
        }
        else {
            $db->updateOrInsertData('aze',{
                    cbuser => $self->user->userId,
                    start => strftime('%Y-%m-%d %H:%M:%S',localtime(time)),
                    lastmod => strftime('%Y-%m-%d %H:%M:%S',localtime(time))
                        . ' - RecStart by ' . $self->user->userInfo->{cbuser_login}
                }
            );
        }
        return {
            action => 'reload'
        }
    };

    my @actions = (
        {
            action => 'refresh',
            interval => 1
        }
    );

        if ( $self->user and $self->user->may('record') ) {
            push @actions, {
                label => trm('Zeit Erfassen'),
                key => 'record',
                action => 'submit',
                handler => $recordTime
            };
        }

        if (not $self->user or $self->user->may('admin') ) {
            push @actions, {
                label => trm('Neu'),
                action => 'popup',
                name => 'azeFormAdd',
                popupTitle => trm('Neuer Zeit Eintrag'),
                backend => {
                    plugin => 'AzeForm',
                    config => {
                        type => 'add'
                    }
                }
            },
            {
                label => trm('Bearbeiten'),
                action => 'popup',
                name => 'azeFormEdit',
                popupTitle => trm('Zeiteintrag Bearbeiten'),
                backend => {
                    plugin => 'AzeForm',
                    config => {
                        type => 'edit'
                    }
                }
            },
            {
                label => trm('Löschen'),
                action => 'submitVerify',
                question => trm('Diesen Eintrag wirklich löschen?'),
                key => 'delete',
                handler => sub {
                    my $args = shift;
                    my $id = $args->{selection}{aze_id};
                    die mkerror(4992,"Bitte einen Eintrag wählen")
                        if not $id;
                    $self->user->db->deleteData('aze',$id);
                    return {
                        action => 'reload'
                    };
                }
            };
    }
    return \@actions;
};

sub getTableRowCount {
    my $self = shift;
    my $args = shift;
    my $db = $self->user->mojoSqlDb;
    if ($self->user->may('admin')){
        return ($db->dbh->selectrow_array("SELECT count(aze_id) FROM aze"))[0];
    }
    else {
        return ($db->dbh->selectrow_array(
            "SELECT count(aze_id) FROM aze WHERE aze_cbuser = ? ",{},$self->user->userId)
        )[0];

    }
}

sub getTableData {
    my $self = shift;
    my $args = shift;
    my $db = $self->user->mojoSqlDb;
    my $SORT ='ORDER by aze_id DESC';
    if ($args->{sortColumn}){
        $SORT = 'ORDER BY '.$db->dbh->quote_identifier($args->{sortColumn});
        $SORT .= $args->{sortDesc} ? ' DESC' : ' ASC';
    }
    my $WHERE = '';
    if (! $self->user->may('admin')){
        $WHERE = ' WHERE aze_cbuser = ' . $db->dbh->quote($self->user->userId);
    }

    return $db->dbh->selectall_arrayref(<<"SQL_END",{Slice => {}}, $args->{lastRow}-$args->{firstRow}+1,$args->{firstRow});
SELECT aze_id,aze_cbuser,aze_start,aze_duration,aze_lastmod,aze_note,
    cbuser_login,
    date(aze_start) AS c_date,
    time(aze_start) AS c_begin,
    time(aze_start,'+' || aze_duration || ' seconds') AS c_end,
    time(aze_duration,'unixepoch') as c_total
FROM aze JOIN cbuser ON aze_cbuser = cbuser_id
$WHERE
$SORT
LIMIT ? OFFSET ?
SQL_END
}

1;
__END__

=head1 COPYRIGHT

Copyright (c) 2015 by Tobias Oetiker. All rights reserved.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2015-34-08/27/15 to 0.0 first version

=cut
